seismic/xcorqc/utils.py

```diff
@@ -0,0 +1,200 @@
+                st = None
+                raise
+            # end try
+
+            bazr = np.radians(baz)
+            tdata = - ste[0].data * np.cos(bazr) + stn[0].data * np.sin(bazr)
+
+            stt = ste.copy()
+            stt[0].data = tdata
+            #stt[0].stats.channel = '00T'
+        # end for
+    # end if
+
+    return stt
+# end func
+
+def get_stream(fds, net, sta, cha, start_time, end_time,
+               baz=None, automerge=False, trace_count_threshold=200,
+               logger=None, verbose=1):
+
+    if (cha == '00T'): return _get_stream_00T(fds, net, sta, cha, start_time, end_time,
+                                              baz=baz, automerge=automerge,
+                                              trace_count_threshold=trace_count_threshold,
+                                              logger=logger, verbose=verbose)
+    st = None
+    stations = fds.get_stations(start_time, end_time, network=net, station=sta)
+    for codes in stations:
+        if (cha != codes[3]): continue
+        st = fds.get_waveforms(codes[0], codes[1], codes[2], codes[3], start_time,
+                               end_time, automerge=automerge, trace_count_threshold=trace_count_threshold)
+
+        if (len(st) == 0): continue
+        drop_bogus_traces(st)
+
+        if (verbose > 2):
+            if logger: logger.debug('\t\tData Gaps:')
+            st.print_gaps()  # output sent to stdout; fix this
+            print ("\n")
+        # end if
+        # Merge station data. Note that we don't want to fill gaps; the
+        # default merge() operation creates masked numpy arrays, which we can use
+        # to detect and ignore windows that have gaps in their data.
+        try:
+            st.merge()
+        except Exception as e:
+            if logger: logger.warning('\tFailed to merge traces..')
+            st = None
+            raise
+        # end try
+    # end for
+
+    return st
+# end func
+
+class ProgressTracker:
+    def __init__(self, output_folder, restart_mode=False):
+        self.output_folder = output_folder
+        self.restart_mode = restart_mode
+
+        self.comm = MPI.COMM_WORLD
+        self.nproc = self.comm.Get_size()
+        self.rank = self.comm.Get_rank()
+
+        self.prev_progress = 0 # progress from a previous run
+        self.progress = 0
+        self.proc_fn = os.path.join(output_folder, 'prog.%d.txt' % (self.rank))
+
+        if(self.restart_mode):
+            if(not os.path.exists(self.proc_fn)):
+                raise Exception('Progress file (%s) not found'%(self.proc_fn))
+            # end if
+
+            self.prev_progress = int(open(self.proc_fn).read())
+        # end if
+    # end func
+
+    def increment(self):
+        self.progress += 1
+        if(self.restart_mode and (self.prev_progress > 0) and (self.progress < self.prev_progress)):
+            return False
+        else:
+            tmpfn = self.proc_fn + '.tmp'
+            f = open(tmpfn, 'w+')
```

Comment:
Consider using context management in the form `with open(tmpfn, 'w+') as f:` to ensure file is closed in case of exception. This would ensure that any successfully written bytes which have not been flushed are properly written to file.